fiducial_vlam/src/vloc_node.cpp

```diff
@@ -325,30 +325,13 @@ namespace fiducial_vlam
       pwc.covariance[35] = 2e-3;
     }
   };
-}
-
-// ==============================================================================
-// main()
-// ==============================================================================
-
-int main(int argc, char **argv)
-{
-  // Force flush of the stdout buffer
-  setvbuf(stdout, nullptr, _IONBF, BUFSIZ);
 
-  // Init ROS
-  rclcpp::init(argc, argv);
-
-  // Create node
-  auto node = std::make_shared<fiducial_vlam::VlocNode>();
-  auto result = rcutils_logging_set_logger_level(node->get_logger().get_name(), RCUTILS_LOG_SEVERITY_INFO);
-  (void) result;
-
-  // Spin until rclcpp::ok() returns false
-  rclcpp::spin(node);
-
-  // Shut down ROS
-  rclcpp::shutdown();
-
-  return 0;
+  std::shared_ptr<rclcpp::Node> vloc_node_factory(const rclcpp::NodeOptions &options)
```

Comment:
I like this pattern, but I'm wondering why you're using it here. Is the idea that there will be other constructor arguments that _main won't need to know about?

Reply:
No real reason except that I like to hide as much as possible and this hides about everything possible. I would prefer to make the result a unique_ptr<>.  It is easy to go from unique_ptr<> to shared_ptr<> but impossible to go the other way.

Reply:
LGTM

Reply:
Now that I think about it, a unique_ptr<> wouldn't be possible in this case. unique_ptr<> requires the class destructor (and therefor the class definition) to be visible when the object is being destructed. I believe this is the case because unique_ptr stores a compile-time indication of how the instance memory was allocated and therefor how it should be deallocated. While shared_ptr has a runtime pointer to a deallocate routine. Besides reference counting, this is another reason shared_ptr is more expensive that unique_ptr (But shared_ptr is still lightweight and cool). unique_ptr<> does work well as the pimpl in the pimpl pattern.